### PR TITLE
fix: respect Snyk Code exclusions on paths with special characters

### PR DESCRIPTION
**Pull request by danskmt**
### Description

Snyk Code file exclusions from `.snyk`, `.gitignore`, and `.dcignore` were silently ignored when the project's absolute path contained regex/glob metacharacters — most commonly parentheses, e.g. `C:\Users\...\OneDrive - Company (Tenant)\...` or `C:\Program Files (x86)\...`.

`parseIgnoreRuleToGlobs` prepends the absolute base path to every rule, and the resulting glob is compiled into a regex by `sabhiram/go-gitignore`. The base path is not pattern data, but it was treated as such: `escapeSpecialGlobChars` only escaped `$`, so characters like `(`, `)`, `+`, `{`, `}`, `^`, `|`, and the glob wildcards `*`, `[`, `]` in the path were interpreted specially and matching silently failed — so nothing was excluded and directories like `node_modules` got scanned in full.

`buildGlob` now escapes the literal base path separately from the glob parts, so path characters always match literally while the rule/`**` parts keep their wildcard meaning. `?` and `.` are intentionally not escaped: `go-gitignore` already treats `?` as literal and escapes `.` itself.

Reproduced independently of OS: a project on a plain path excludes `node_modules` (~16 files bundled), while the same project under a `(...)` path did not (~28.9k files, matching a no-exclusion scan). After the fix, both exclude correctly.

Ticket: CLI-1648

### Checklist

- [x] Tests added and all succeed (`make test`)
- [x] Regenerated mocks, etc. (`make generate`)
- [x] Linted (`make lint`)
- [x] Test your changes work for the CLI
  1. Clone / pull the latest [CLI](https://github.com/snyk/cli) main.
  2. Run `go get github.com/snyk/go-application-framework@YOUR_LATEST_GAF_COMMIT` in the `cliv2` directory.
      - Tip: for local testing, you can uncomment the line near the bottom of the CLI's [`go.mod`](https://github.com/snyk/cli/blob/main/cliv2/go.mod) to point to your local GAF code.
  3. Run `go mod tidy` in the `cliv2` directory.
  4. Run the CLI tests and do any required manual testing.
  5. Open a PR in the CLI repo **now** with the `go.mod` and `go.sum` changes.
  - Once this PR is merged, repeat these steps, but pointing to the latest GAF commit on main and update your CLI PR.

https://github.com/snyk/cli/pull/6980

---

> [!NOTE]
> **Medium Risk**
> Changes which files are included in Snyk Code scans; incorrect matching could over- or under-scan, though behavior is covered by new scenario tests.
> 
> **Overview**
> Fixes **silent failure of Snyk Code file exclusions** (`.gitignore`, `.snyk`, `.dcignore`) when the scan root’s absolute path contains characters that `go-gitignore` treats as regex/glob syntax—common on Windows paths like `Program Files (x86)` or OneDrive folder names with parentheses.
> 
> `parseIgnoreRuleToGlobs` no longer only escapes `$` on the full joined pattern. It uses **`buildGlob`** to escape the **literal base directory** aggressively (`*`, `[`, `]`, `^`, `\`, plus regex metacharacters) while still allowing **gitignore rule syntax** such as negated classes (`cache[^S]`) on the rule suffix. A conservative fallback applies when a rule’s `../` climbs above `baseDir`.
> 
> Adds broad tests for meta-character paths, nested ignore files, directory rules, and the `../` glob edge case.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 07a51b4f473f5c5bde3fe4bdb7bea48d2ade7167. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pkg/utils/file_filter.go [hunk 1]
```diff
@@ -326,23 +326,62 @@ func parseIgnoreFile(content []byte, filePath string) []string {
...
+	'|': true,
+	'{': true,
+	'}': true,
+}
+
+// pathMetaCharsToEscape extends regexMetaCharsToEscape with the glob wildcards "*", "[" and
+// "]" for escaping the literal base path (which is not a pattern), e.g. "Program Files (x86)"
+// or a folder named "a*b". "?" and "." are omitted: the ignore matcher already treats them
+// literally / escapes them itself.
+var pathMetaCharsToEscape = map[byte]bool{
```
Comment (basti-snyk) [inline pkg/utils/file_filter.go]: Suggestion: `pathMetaCharsToEscape` duplicates all 8 entries of `regexMetaCharsToEscape` verbatim, then adds `* [ ]`. The two maps can drift — a future regex metachar added to one but not the other would silently change base-path escaping. Consider deriving `pathMetaCharsToEscape` from `regexMetaCharsToEscape` (copy, then add the three glob chars) so the shared set lives in one place.

— AI review

File: pkg/utils/file_filter.go [hunk 2]
```diff
@@ -326,23 +326,62 @@ func parseIgnoreFile(content []byte, filePath string) []string {
...
+	'^': true,
+	'|': true,
+	'{': true,
+	'}': true,
+	'*': true,
+	'[': true,
+	']': true,
+}
+
+func escapeChars(s string, charsToEscape map[byte]bool) string {
```
Comment (basti-snyk) [inline pkg/utils/file_filter.go:355]: Should fix: a literal backslash `\` in a directory name is not escaped. On Linux/macOS `\` is a legal filename char and `filepath.ToSlash` is a no-op, so a dir named `a\b` reaches the matcher unescaped; `sabhiram/go-gitignore` + Go RE2 then read `\b` as a word boundary (or an invalid `\c` fails to compile → the rule is silently dropped), losing the exclusion. This is the same bug class this PR fixes, left uncovered — empirically confirmed that `MatchesPath` returns false for a path under `a\b`. Consider adding backslash to `pathMetaCharsToEscape` (ordering is safe since `escapeChars` only prepends). Narrow non-Windows edge, noting because it belongs in this function.

— AI review

File: pkg/utils/file_filter.go [hunk 3]
```diff
@@ -326,23 +326,62 @@ func parseIgnoreFile(content []byte, filePath string) []string {
...
+// characters are never interpreted as patterns) while the glob parts keep their wildcards.
+func buildGlob(prefix, baseDir string, parts ...string) string {
+	base := filepath.ToSlash(filepath.Clean(baseDir))
+	joined := filepath.ToSlash(filepath.Join(append([]string{baseDir}, parts...)...))
+	if rest, ok := strings.CutPrefix(joined, base); ok {
+		return prefix + escapeChars(base, pathMetaCharsToEscape) + escapeChars(rest, regexMetaCharsToEscape)
+	}
+	// Fallback: base is not a prefix of the joined path (e.g. a rule with "../" escaped above
+	// it). Escape conservatively as a glob so at least regex metacharacters are handled.
+	return escapeChars(prefix+joined, regexMetaCharsToEscape)
```
Comment (basti-snyk) [inline pkg/utils/file_filter.go:377]: Suggestion: the fallback branch escapes with `regexMetaCharsToEscape`, which omits `*`, `[`, `]`. A base path containing `a*b` or `a[b]c` combined with a rule that climbs above `baseDir` via `../` would re-expose the exact metachar-in-path bug this PR fixes. Reachability is very low (needs `../`-heavy rules surviving `filepath.Join` cleaning) and the branch has no test coverage. Consider escaping the fallback with `pathMetaCharsToEscape` for consistency, or add a climbing-rule test case.

— AI review

**Comment by robertolopezlopez** [comment]
suggestion (long term fix): you may leverage `utils.ToRelativeUnixPath()` (from `path.go`) . This way you can drop the `go-gitignore` requirement, which is enforcing regex (the root cause of this bug). 

For example in `GetFilteredFiles()`:

change:

`if !globPatternMatcher.MatchesPath(f) {` (line 131)

to:

```
rel, err := ToRelativeUnixPath(fw.path, f)
if err != nil { /* handle the error */ }
if !globPatternMatcher.MatchesPath(rel) {
```

**Comment by danskmt** [comment]
> suggestion (long term fix): you may leverage `utils.ToRelativeUnixPath()` (from `path.go`) . This way you can drop the `go-gitignore` requirement, which is enforcing regex (the root cause of this bug).
> 
> For example in `GetFilteredFiles()`:
> 
> change:
> 
> `if !globPatternMatcher.MatchesPath(f) {` (line 131)
> 
> to:
> 
> ```
> rel, err := ToRelativeUnixPath(fw.path, f)
> if err != nil { /* handle the error */ }
> if !globPatternMatcher.MatchesPath(rel) {
> ```

I added some more test cases where using this strategy would fail (nested patterns)

File: pkg/utils/file_filter.go [hunk 4]
```diff
@@ -326,23 +327,56 @@ func parseIgnoreFile(content []byte, filePath string) []string {
 	return ignores
 }
 
-// escapeSpecialGlobChars escapes special characters that should be treated literally in glob patterns.
-// Special Characters to escape: $
-func escapeSpecialGlobChars(rule string) string {
+// regexMetaCharsToEscape are regex metacharacters that are not glob syntax; escaped in ignore
+// rules so they match literally (e.g. "$" in a rule). "^", "[" and "]" are deliberately not
+// included: they are valid character-class syntax in gitignore rules (e.g. "foo[^x]").
+var regexMetaCharsToEscape = map[byte]bool{
```
Comment (PeterSchafer) [inline pkg/utils/file_filter.go:333]: Suggestion: Can we use this to safely escape the string?
[regexp package - regexp - Go Packages](https://pkg.go.dev/regexp#QuoteMeta)

**Review by robertolopezlopez — COMMENTED** [review]
## Review summary

**Verdict: Approve** — correct root-cause fix with strong test coverage; one low-reachability edge case worth acknowledging.

### Root cause & approach
The diagnosis is right: `parseIgnoreRuleToGlobs` prepends the absolute scan root into every glob, and `sabhiram/go-gitignore` compiles that string to regex. Path characters like `( ) * [ ] ^` were treated as pattern syntax, so exclusions silently failed — matching CLI-1648 / customer paths (`OneDrive - Foobar (Team1)`, `Program Files (x86)`).

`buildGlob` is the right split:
- **Base path** → `pathMetaCharsToEscape` (literal filesystem segment)
- **Rule suffix** → `regexMetaCharsToEscape` only, deliberately **not** escaping `^`, `[`, `]` so gitignore character classes (`cache[^S]`) keep working

The `da4ae565` follow-up commit that preserved negated classes was necessary — good catch.

### What looks good
- `maps.Clone(regexMetaCharsToEscape)` for the path map avoids drift between the two sets (addresses earlier review feedback).
- Backslash included in `pathMetaCharsToEscape` — relevant on Unix paths like `a\b`.
- Broad tests: per-metachar path matrix (with Windows-illegal skips), table-driven rule scenarios (nested ignores, scoped rules, directory rules), and a unit test for the `../` fallback glob shape.
- CI green including **Windows tests** — important for this bug class.
- CLI companion PR linked ([snyk/cli#6980](https://github.com/snyk/cli/pull/6980)).

### Non-blocking nits

1. **Fallback branch** (`buildGlob` when `CutPrefix` fails): still escapes with `regexMetaCharsToEscape`, so `*`, `[`, `]` in ancestor path segments could theoretically slip through if a `../`-heavy rule climbs above `baseDir` while the resolved path still contains filesystem metachars (e.g. `/tmp/a*b/...`). Reachability is very low and the primary customer path is covered; acceptable to ship, but a one-line comment documenting the limitation would help future readers.

2. **`regexp.QuoteMeta`** (per Peter's comment): not suitable for the full joined string — it would break rule character classes. The per-segment split here is the correct trade-off.

### Post-merge
- Refresh [CLI #6980](https://github.com/snyk/cli/pull/6980) to the merged GAF commit (currently pins `2b4c6a0dbdab`, head is `07a51b4f473f`).

### Risk
Medium impact surface (which files enter the Code bundle) but change is narrowly scoped and fail-safe direction is *more* exclusions applied, not fewer.

File: pkg/utils/file_filter.go (hunk 4 above)
Comment (danskmt) [reply pkg/utils/file_filter.go:333]: It seems we cannot - the LLM response here is a good one to see why not:
image

Comment (robertolopezlopez) [reply pkg/utils/file_filter.go:333]: I asked cursor to craft this table:

| Target | QuoteMeta | Result |
| :--- | :--- | :--- |
| Path segment my (dir) | my \(dir\) | OK-ish |
| Path segment v1.0 | v1\.0 | Breaks — go-gitignore re-escapes . |
| Path segment a?b | a\?b | Breaks — double-escape on ? |
| Rule suffix /node_modules/ | \\/node_modules/\\ | Wildcards dead |
| Rule cache[^S] | cache\[\^S\] | Class dead |
| Rule price$tag | price\$tag | Works (same as escapeSpecialGlobChars) |

Comment (danskmt) [reply pkg/utils/file_filter.go:333]: @PeterSchafer and @robertolopezlopez it seems we have many gaps coming from go-gitignore, I added more info in a comment in [CLI-1410](https://snyksec.atlassian.net/browse/CLI-1410)

**Comment by robertolopezlopez** [comment]
> > suggestion (long term fix): you may leverage `utils.ToRelativeUnixPath()` (from `path.go`) . This way you can drop the `go-gitignore` requirement, which is enforcing regex (the root cause of this bug).
> > For example in `GetFilteredFiles()`:
> > change:
> > `if !globPatternMatcher.MatchesPath(f) {` (line 131)
> > to:
> > ```
> > rel, err := ToRelativeUnixPath(fw.path, f)
> > if err != nil { /* handle the error */ }
> > if !globPatternMatcher.MatchesPath(rel) {
> > ```
> 
> I added some more test cases where using this strategy would fail (nested patterns)

@danskmt 

In my PR those tests pass ;-) https://github.com/snyk/go-application-framework/pull/654

File: pkg/utils/file_filter.go (hunk 4 above)
Comment (PeterSchafer) [reply pkg/utils/file_filter.go:333]: @danskmt sorry this might be a misunderstanding, I'm not asking to escape all characters from the whole pattern, just from the base path. This is what is not a pattern

Comment (PeterSchafer) [reply pkg/utils/file_filter.go:333]: and sorry AI, I don't agree 😊

**Review by PeterSchafer — CHANGES_REQUESTED** [review]

File: pkg/utils/file_filter.go (hunk 4 above)
Comment (danskmt) [reply pkg/utils/file_filter.go:333]: @PeterSchafer you mean in line 373 to have this?
```go
return prefix + regexp.QuoteMeta(base) + escapeChars(rest, regexMetaCharsToEscape)
```

**Comment by danskmt** [comment]
Closing this PR in favor of https://github.com/snyk/go-application-framework/pull/656
